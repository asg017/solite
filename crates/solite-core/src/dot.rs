use std::{
    io::{BufRead, BufReader}, path::PathBuf, sync::mpsc::Receiver
};

use solite_stdlib::solite_stdlib_init;
use thiserror::Error;

use crate::{exporter::write_output, sqlite::{Connection, Statement}, Runtime};
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Error, Debug, PartialEq)]
pub enum ParseDotError {
    #[error("Unknown command '{0}'")]
    UnknownCommand(String),
    #[error("Invalid argument: {0}")]
    InvalidArgument(String),

    #[error("{0}")]
    Generic(String),
}

#[derive(Serialize, Debug, PartialEq)]
pub struct PrintCommand {
    pub message: String,
}
impl PrintCommand {
    pub fn execute(&self) {
        println!("{}", self.message);
    }
}
#[derive(Serialize, Debug, PartialEq)]
pub struct ShellCommand {
    pub command: String,
}
impl ShellCommand {
    pub fn execute(&self) -> Receiver<String> {
        let (tx, mut rx) = std::sync::mpsc::channel::<String>();
        let command = self.command.clone();
        std::thread::spawn(move || {
            let mut child = std::process::Command::new("sh")
                .arg("-c")
                .arg(command)
                .stdout(std::process::Stdio::piped())
                .spawn()
                .unwrap();

            let stdout = child.stdout.take().expect("Failed to capture stdout");
            let reader = BufReader::with_capacity(1, stdout); //new(stdout);
            let mut lines = reader.lines();
            while let Some(Ok(line)) = lines.next() {
                tx.send(line).unwrap();
            }
            let _ = child.wait();
        });
        return rx;
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub struct TablesCommand {}
impl TablesCommand {
    pub fn execute(&self, runtime: &Runtime) {
        let stmt = runtime
            .connection
            .prepare(
                r#"
                select name
                from pragma_table_list
                where "schema" = 'main'
                  and type in ('table', 'view')
                  and name not like 'sqlite_%'
                order by name;
                "#,
            )
            .unwrap()
            .1
            .unwrap();
        let mut tables = vec![];
        while let Ok(Some(row)) = stmt.next() {
            tables.push(row.get(0).unwrap().as_str().to_owned());
        }
        for table in tables {
            println!("{table}")
        }
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub struct OpenCommand {
    pub path: String,
}
impl OpenCommand {
    pub fn execute(&self, runtime: &mut Runtime) {
        runtime.connection = Connection::open(&self.path).unwrap();
        unsafe {
            solite_stdlib_init(
                runtime.connection.db(),
                std::ptr::null_mut(),
                std::ptr::null_mut(),
            );
        }
    }
}

#[derive(Serialize, Debug)]
pub struct ExportCommand {
    pub target: PathBuf,
    pub statement: Statement,
    pub rest_length: usize,
}

impl ExportCommand {
    pub fn new(args: String, runtime: &mut Runtime, rest: &str) -> Result<Self, ParseDotError> {
      match runtime.prepare_with_parameters(rest) {
        Ok((rest2, Some(stmt))) => {
          Ok(Self {
            target: PathBuf::from(args),
            statement: stmt,
            // TODO: suspicious
            rest_length: rest2.unwrap_or(rest.len())
          })
        },
        _ => todo!(),
    }
  }
    pub fn execute(&mut self) -> anyhow::Result<()> {
        let output = crate::exporter::output_from_path(&self.target)
            .map_err(|e| ParseDotError::Generic(e.to_string()))?;
        let format = crate::exporter::format_from_path(&self.target)
            .unwrap();
          write_output(&mut self.statement, output, format).unwrap();
        Ok(())
  }

}

#[derive(Serialize, Debug)]
pub struct VegaLiteCommand {
    pub statement: Statement,
    pub rest_length: usize,
}

impl VegaLiteCommand {
    pub fn new(args: String, runtime: &mut Runtime, rest: &str) -> Result<Self, ParseDotError> {
      match runtime.prepare_with_parameters(rest) {
        Ok((rest2, Some(stmt))) => {
          Ok(Self {
            statement: stmt,
            // TODO: suspicious
            rest_length: rest2.unwrap_or(rest.len())
          })
        },
        _ => todo!(),
    }
  }
    pub fn execute(&mut self) -> anyhow::Result<serde_json::Map<String, serde_json::Value>> {
        let columns = self.statement.column_names().unwrap();
        let mark = match columns[0].as_str() { 
          "bar" => "bar",
          _ => todo!(),
        };
        let mut data = vec![];
        loop {
          match self.statement.nextx() {
            Ok(Some(row)) => {
              let x = match row.value_at(1).value {
                crate::sqlite::ValueRefXValue::Blob(_) => serde_json::Value::Null,
                crate::sqlite::ValueRefXValue::Int(value) => serde_json::Value::Number(value.into()),
                crate::sqlite::ValueRefXValue::Double(value) => serde_json::Number::from_f64(value).map(serde_json::Value::Number).unwrap_or(serde_json::Value::Null),
                crate::sqlite::ValueRefXValue::Text(value) => serde_json::Value::Null,
                crate::sqlite::ValueRefXValue::Null => serde_json::Value::Null,
              };
              let y = match row.value_at(2).value {
                crate::sqlite::ValueRefXValue::Blob(_) => serde_json::Value::Null,
                crate::sqlite::ValueRefXValue::Int(value) => serde_json::Value::Number(value.into()),
                crate::sqlite::ValueRefXValue::Double(value) => serde_json::Number::from_f64(value).map(serde_json::Value::Number).unwrap_or(serde_json::Value::Null),
                crate::sqlite::ValueRefXValue::Text(value) => serde_json::Value::Null,
                crate::sqlite::ValueRefXValue::Null => serde_json::Value::Null,
              };
              data.push(serde_json::json!({
                "x": x,
                "y": y,
              }));
            },
            Ok(None) => break,
            Err(_) => todo!(),
            }
        }

        let data = serde_json::json!({
          "$schema": "https://vega.github.io/schema/vega-lite/v6.json",
          "description": "A simple bar chart with embedded data.",
          "data": {
            "values": data,
          },
          "mark": mark,
          "encoding": {
            "x": {"field": "x", "type": "quantitative"},
            "y": {"field": "y", "type": "quantitative"}
          }
        });
        Ok(data.as_object().cloned().unwrap())
  }

}

#[derive(Serialize, Debug, PartialEq)]
pub struct LoadCommand {
    pub path: String,
    pub entrypoint: Option<String>,
    pub is_uv: bool,
}

pub enum LoadCommandSource {
    Path(String),
    Uv { directory: String, package: String },
}

impl LoadCommand {
    pub fn new(args: String) -> Self {
        let (args, is_uv) = match args.strip_prefix("uv:") {
            Some(args) => (args, true),
            None => (args.as_str(), false),
        };

        let (path, entrypoint) = match args.split_once(' ') {
            Some((path, entrypoint)) => (path.to_string(), Some(entrypoint.trim().to_string())),
            None => (args.to_owned(), None),
        };
        Self {
            path,
            entrypoint,
            is_uv,
        }
    }
    pub fn execute(&self, connection: &mut Connection) -> anyhow::Result<LoadCommandSource> {
        if self.is_uv {
            crate::load_uv::load(connection, &self.path, &self.entrypoint).map(|path| {
                LoadCommandSource::Uv {
                    directory: path,
                    package: self.path.clone(),
                }
            })
        } else {
            connection
                .load_extension(&self.path, &self.entrypoint)
                .map(|_| LoadCommandSource::Path(self.path.clone()))
        }
    }
}

#[derive(Serialize, Debug, PartialEq)]
pub enum ParameterCommand {
    Set { key: String, value: String },
    Unset(String),
    List,
    Clear,
}

fn parse_parameter(line: String) -> ParameterCommand {
    match line.trim_end().split_once(' ') {
        Some((word, rest)) => match word {
            "set" => {
                let (k, v) = rest.split_once(' ').unwrap();
                ParameterCommand::Set {
                    key: k.to_owned(),
                    value: v.to_owned(),
                }
            }
            "unset" => ParameterCommand::Unset(rest.to_owned()),
            _ => todo!(),
        },
        None => match line.trim_end() {
            "list" => ParameterCommand::List,
            "clear" => ParameterCommand::Clear,
            _ => todo!(),
        },
    }
}
#[derive(Serialize, Debug)]
pub enum DotCommand {
    /*  introspection  */
    //Databases,
    Tables(TablesCommand),
    //Indexes,
    //Schema,

    /*  docs  */
    //Help,
    //Docs,

    /*  runtime  */
    /// .run file.sql -p a=1 -p b=2
    //Run,
    /// or .debugger?
    //Breakpoint,

    //Quit,
    //Exit,

    /// switches to different DB connection
    /// usage: .open file.db
    Open(OpenCommand),
    Load(LoadCommand),

    /// TODO sqlpkg/spm support
    //Load,
    Print(PrintCommand),

    Shell(ShellCommand),
    /// usage: .param set name 'alex garcia'
    Parameter(ParameterCommand),
    /// usage: .bail on/off
    //Bail,

    /// usage: .timer on/off
    Timer(bool),
    //
    //Mode,
    Export(ExportCommand),
    Vegalite(VegaLiteCommand),
}

fn parse_bool(s: String) -> Result<bool, String> {
    match s.to_lowercase().as_str() {
        "yes" | "y" | "on" => Ok(true),
        "no" | "n" | "off" => Ok(false),
        _ => Err(format!("Not a boolean value: {}", s)),
    }
}
pub fn parse_dot<S: Into<String>>(command: S, args: S, rest: &str, runtime: &mut Runtime) -> Result<DotCommand, ParseDotError> {
    let command = command.into();
    let args = args.into();
    match command.to_lowercase().as_str() {
        "print" => Ok(DotCommand::Print(PrintCommand { message: args })),
        "sh" => Ok(DotCommand::Shell(ShellCommand { command: args })),
        "tables" => Ok(DotCommand::Tables(TablesCommand {})),
        "open" => Ok(DotCommand::Open(OpenCommand { path: args })),
        "export" => Ok(DotCommand::Export(ExportCommand::new(args, runtime, rest)?)),
        "vl" | "vegalite" => Ok(DotCommand::Vegalite(VegaLiteCommand::new(args, runtime, rest)?)),
        "load" => Ok(DotCommand::Load(LoadCommand::new(args))),
        "timer" => Ok(DotCommand::Timer(
            parse_bool(args).map_err(ParseDotError::InvalidArgument)?,
        )),
        "param" | "parameter" => Ok(DotCommand::Parameter(parse_parameter(args))),
        _ => Err(ParseDotError::UnknownCommand(command)),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_dot() {
    }
}