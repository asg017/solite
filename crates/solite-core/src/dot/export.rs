use serde::Serialize;
use crate::{
    exporter::write_output,
    sqlite::{OwnedValue,Statement},
    Runtime,
    ParseDotError
};
use std::path::PathBuf;
use regex::{Captures, Regex};


#[derive(Serialize, Debug)]
pub struct ExportCommand {
    pub target: PathBuf,
    pub statement: Statement,
    pub rest_length: usize,
}

impl ExportCommand {
    pub fn new(args: String, runtime: &mut Runtime, rest: &str) -> Result<Self, ParseDotError> {
        match runtime.prepare_with_parameters(rest) {
            Ok((rest2, Some(stmt))) => {
                let re = Regex::new(r":[\w]+").unwrap();
                let target = re.replace_all(&args, |cap:&Captures| {
                    let param_name = &cap[0].strip_prefix(":").unwrap(); 
                    match runtime.lookup_parameter(param_name) {
                        Some(value) => match value {
                          OwnedValue::Text(text) => {
                            std::str::from_utf8(&text).unwrap().to_string()
                          },
                          _ => "".to_owned(),
                        },
                        None => {
                            "".to_owned()
                        }
                    }
                });
                Ok(Self {
                    target: PathBuf::from(target.to_string()),
                    statement: stmt,
                    // TODO: suspicious
                    rest_length: rest2.unwrap_or(rest.len()),
                })
            }
            _ => todo!(),
        }
    }
    pub fn execute(&mut self) -> anyhow::Result<()> {
        let output = crate::exporter::output_from_path(&self.target)
            .map_err(|e| ParseDotError::Generic(e.to_string()))?;
        let format = crate::exporter::format_from_path(&self.target).unwrap();
        write_output(&mut self.statement, output, format).unwrap();
        Ok(())
    }
}
