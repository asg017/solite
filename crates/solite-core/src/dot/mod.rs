mod bench;
mod load;
mod export;
mod open;
mod param;
mod print;
mod sh;
mod tables;
mod timer;
mod vegalite;

pub use crate::dot::{
  bench::BenchCommand,
  load::LoadCommand,
  sh::ShellCommand,
  print::PrintCommand,
  tables::TablesCommand,
  open::OpenCommand,
  export::ExportCommand,
  vegalite::VegaLiteCommand,
  param::ParameterCommand,
};
pub use load::LoadCommandSource;

use param::parse_parameter;
use thiserror::Error;

use crate::Runtime;
use serde::{Deserialize, Serialize};

#[derive(Serialize, Deserialize, Error, Debug, PartialEq)]
pub enum ParseDotError {
    #[error("Unknown command '{0}'")]
    UnknownCommand(String),
    #[error("Invalid argument: {0}")]
    InvalidArgument(String),

    #[error("{0}")]
    Generic(String),
}





#[derive(Serialize, Debug)]
pub enum DotCommand {
    /*  introspection  */
    //Databases,
    Tables(TablesCommand),
    //Indexes,
    //Schema,

    /*  docs  */
    //Help,
    //Docs,

    /*  runtime  */
    /// .run file.sql -p a=1 -p b=2
    //Run,
    /// or .debugger?
    //Breakpoint,

    //Quit,
    //Exit,

    /// switches to different DB connection
    /// usage: .open file.db
    Open(OpenCommand),
    Load(LoadCommand),

    /// TODO sqlpkg/spm support
    //Load,
    Print(PrintCommand),

    Shell(ShellCommand),
    /// usage: .param set name 'alex garcia'
    Parameter(ParameterCommand),
    /// usage: .bail on/off
    //Bail,

    /// usage: .timer on/off
    Timer(bool),
    //
    //Mode,
    Export(ExportCommand),
    Vegalite(VegaLiteCommand),
    Bench(BenchCommand),
}

fn parse_bool(s: String) -> Result<bool, String> {
    match s.to_lowercase().as_str() {
        "yes" | "y" | "on" => Ok(true),
        "no" | "n" | "off" => Ok(false),
        _ => Err(format!("Not a boolean value: {}", s)),
    }
}
pub fn parse_dot<S: Into<String>>(
    command: S,
    args: S,
    rest: &str,
    runtime: &mut Runtime,
) -> Result<DotCommand, ParseDotError> {
    let command = command.into();
    let args = args.into();
    match command.to_lowercase().as_str() {
        "print" => Ok(DotCommand::Print(PrintCommand { message: args })),
        "sh" => Ok(DotCommand::Shell(ShellCommand { command: args })),
        "tables" => Ok(DotCommand::Tables(TablesCommand {})),
        "open" => Ok(DotCommand::Open(OpenCommand { path: args })),
        "export" => Ok(DotCommand::Export(ExportCommand::new(args, runtime, rest)?)),
        "bench" => Ok(DotCommand::Bench(BenchCommand::new(args, runtime, rest)?)),
        "vl" | "vegalite" => Ok(DotCommand::Vegalite(VegaLiteCommand::new(
            args, runtime, rest,
        )?)),
        "load" => Ok(DotCommand::Load(LoadCommand::new(args))),
        "timer" => Ok(DotCommand::Timer(
            parse_bool(args).map_err(ParseDotError::InvalidArgument)?,
        )),
        "param" | "parameter" => Ok(DotCommand::Parameter(parse_parameter(args))),
        _ => Err(ParseDotError::UnknownCommand(command)),
    }
}

#[cfg(test)]
mod tests {
    #[test]
    fn test_parse_dot() {}
}
