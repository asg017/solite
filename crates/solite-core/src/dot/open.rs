use serde::Serialize;
use crate::{Runtime, Connection};
use solite_stdlib::solite_stdlib_init;

#[derive(Serialize, Debug, PartialEq)]
pub struct OpenCommand {
    pub path: String,
}
impl OpenCommand {
    pub fn execute(&self, runtime: &mut Runtime) {
        runtime.connection = Connection::open(&self.path).unwrap();
        unsafe {
            solite_stdlib_init(
                runtime.connection.db(),
                std::ptr::null_mut(),
                std::ptr::null_mut(),
            );
        }
    }
}