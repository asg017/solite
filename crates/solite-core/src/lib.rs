pub mod dot;
pub mod load_uv;
pub mod replacement_scans;
pub mod sqlite;
pub mod exporter;

use crate::dot::{DotCommand, ShellCommand};
use crate::sqlite::Connection;
use dot::{parse_dot, ParseDotError};
use libsqlite3_sys::{
    sqlite3_db_config, SQLITE_DBCONFIG_DEFENSIVE, SQLITE_DBCONFIG_WRITABLE_SCHEMA,
};
use regex::Regex;
use ropey::Rope;
use serde::{Deserialize, Serialize};
use solite_stdlib::solite_stdlib_init;
use sqlite::{OwnedValue, SQLiteError, Statement};
use std::sync::LazyLock;
use std::{fmt, path::PathBuf};
use thiserror::Error;

static SQL_COMMENT_REGION: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\s*--\s*#region\s+(\w*)").unwrap());
static SQL_COMMENT_ENDREGION: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\s*--\s*#endregion").unwrap());

fn sql_comment_region_name(sql: &str) -> Option<&str> {
    //SQL_COMMENT_REGION.captures_at(sql, 1).map(|x| x)
    SQL_COMMENT_REGION
        .captures(sql)
        .and_then(|captures| captures.get(1).map(|m| m.as_str()))
}

#[derive(Serialize, Deserialize, Error, Debug)]
pub enum StepError {
    #[error("Error preparing SQL statement:")]
    Prepare {
        file_name: String,
        src: String,
        offset: usize,
        error: SQLiteError,
    },
    #[error("Error parsing dot command: {0}")]
    ParseDot(ParseDotError),
}

/// A "block" of commands to run - can be SQL, PRQL(?) or dot commands.
/// Can come from a SQL file ondisk, a Jupyer cell, REPL, etc.
#[derive(Debug)]
pub struct Block {
    // either file name or "[stdin]" or "[script]" or something
    name: String,
    source: BlockSource,
    contents: String,
    rope: Rope,
    offset: usize,
    regions: Vec<String>,
}

#[derive(Debug)]
pub enum BlockSource {
    File(PathBuf),
    Repl,
    JupyerCell,
}

#[derive(Default)]
pub struct State {
    //timer: bool,
    //bail: bool,
}

pub struct Runtime {
    pub connection: Connection,
    stack: Vec<Block>,
    //state: State,
    initialized_sqlite_parameters_table: bool,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct StepReference {
    block_name: String,
    line_number: usize,
    column_number: usize,
    pub region: Vec<String>,
}

impl fmt::Display for StepReference {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}:{}:{}",
            self.block_name, self.line_number, self.column_number
        )
    }
}

#[derive(Serialize, Debug)]
pub enum StepResult {
    SqlStatement { stmt: Statement, raw_sql: String },
    DotCommand(dot::DotCommand),
}

#[derive(Serialize, Debug)]

pub struct Step {
    preamble: Option<String>,
    /// Dot command or SQL
    pub result: StepResult,

    pub reference: StepReference,
}

fn extract_preamble(code: &str) -> (&str, Option<&str>) {
    let codex = advance_through_ignorable(code);
    let preamble_offset = unsafe { codex.as_ptr().offset_from(code.as_ptr()) } as usize;
    if preamble_offset > 0 {
        (codex, Some(&code[0..preamble_offset]))
    } else {
        (code, None)
    }
}

impl Runtime {
    pub fn new(path: Option<String>) -> Self {
        unsafe {
            //libsqlite3_sys::sqlite3_auto_extension(Some(std::mem::transmute(solite_stdlib_init as *const (),)));
        }
        let connection = match path {
            Some(path) => Connection::open(path.as_str()).unwrap(),
            None => Connection::open_in_memory().unwrap(),
        };
        unsafe {
            solite_stdlib_init(connection.db(), std::ptr::null_mut(), std::ptr::null_mut());
        }
        Runtime {
            connection,
            stack: vec![],
            //state: State::default(),
            initialized_sqlite_parameters_table: false,
        }
    }
    pub fn enqueue(&mut self, name: &str, code: &str, source: BlockSource) {
        self.stack.push(Block {
            name: name.to_string(),
            source,
            contents: code.to_string(),
            rope: Rope::from_str(code),
            offset: 0,
            regions: vec![],
        });
    }
    pub fn next_stepx(&mut self) -> Option<Result<Step, StepError>> {
        while let Some(mut block) = self.stack.pop() {
            let regions = block.regions.clone();
            let current = match (&block.contents).get(block.offset..) {
                Some(code) => code,
                None => return None,
            };
            let (code, preamble) = extract_preamble(current);
            if let Some(preamble) = preamble {
                block.offset += code.as_ptr() as usize - preamble.as_ptr() as usize;
                /*if !preamble.is_empty() {
                }*/
                for line in preamble.lines() {
                    if let Some(region) = sql_comment_region_name(line) {
                        block.regions.push(region.to_string());
                    } else if SQL_COMMENT_ENDREGION.is_match(line) {
                        block.regions.pop();
                    }
                }
            }

            if code.starts_with('.') || code.starts_with("!") {
                let end_idx = code.find('\n').unwrap_or(code.len());
                let dot_line = code.get(0..end_idx).unwrap();
                let source = block.name.to_string();
                let rest: &str = code.get(end_idx..).unwrap();
                    

                let cmd = if !code.starts_with('!') {
                    let sep_idx = dot_line.find(' ').unwrap_or(dot_line.len());
                    let dot_command = dot_line.get(1..sep_idx).unwrap().trim().to_string();
                    let dot_args = dot_line.get(sep_idx..).unwrap().trim().to_string();
                    
                    match parse_dot(dot_command, dot_args, rest, self) {
                        Ok(cmd) => {
                          match &cmd {
                            DotCommand::Export(cmd) => {
                              block.offset +=  cmd.rest_length;
                            }
                            DotCommand::Vegalite(cmd) => {
                              block.offset +=  cmd.rest_length;
                            }
                            _ => (),
                          }
                          cmd
                        },
                        Err(err) => {
                            return Some(Err(StepError::ParseDot(err)));
                      }
                    }
                } else {
                    DotCommand::Shell(ShellCommand {
                        command: dot_line.get(1..end_idx).unwrap().trim().to_string(),
                    })
                };

                if !rest.is_empty() {
                      block.offset +=  dot_line.len(); //preamble preamble_offset + end_idx + 1;
                      self.stack.push(block);
                  }

                return Some(Ok(Step {
                    preamble: None,
                    reference: StepReference {
                        block_name: source,
                        // TODO: why hardcode here?
                        line_number: 1,
                        column_number: 1,
                        region: regions,
                    },
                    result: StepResult::DotCommand(cmd),
                }));
            }

            match self.prepare_with_parameters(&code) {
                Ok((rest, Some(stmt))) => {
                    let stmt_offset_idx = block.offset; // + preamble.map_or(0, |p| p.len()) + 1;
                    let block_name = block.name.clone();
                    let line_idx: usize = block.rope.byte_to_line(stmt_offset_idx);
                    let column_idx = stmt_offset_idx - block.rope.line_to_byte(line_idx);
                    let raw_sql = stmt.sql(); //code.to_owned();
                    let preamble_owned = preamble.map(|p| p.to_string());

                    if let Some(rest) = rest {
                        block.offset += rest; // + preamble.map_or(0, |p| p.len()) + 1;
                        self.stack.insert(0, block);
                    }

                    return Some(Ok(Step {
                        preamble: preamble_owned,
                        reference: StepReference {
                            block_name,
                            line_number: line_idx + 1,
                            column_number: column_idx + 1,
                            region: regions,
                        },
                        result: StepResult::SqlStatement { stmt, raw_sql },
                    }));
                }
                Ok((_rest, None)) => {
                    return None;
                }
                Err(error) => {
                    match replacement_scans::replacement_scan(&error, &self.connection) {
                        Some(Ok(stmt)) => {
                            stmt.execute().unwrap();
                            self.stack.push(block);
                        }
                        Some(Err(_)) => todo!(),
                        None => {
                            return Some(Err(StepError::Prepare {
                                error,
                                file_name: block.name,
                                src: block.contents,
                                offset: block.offset,
                            }));
                        }
                    };
                }
            }
        }
        None
    }

    pub fn execute_to_completion(&mut self) -> Result<(), StepError> {
        loop {
            match self.next_stepx() {
                None => return Ok(()),
                Some(Ok(step)) => {
                    match step.result {
                        StepResult::SqlStatement { stmt, .. } => stmt.execute().unwrap(),
                        StepResult::DotCommand(_cmd) => todo!(),
                    };
                    continue;
                }
                Some(Err(err)) => return Err(err),
            }
        }
    }
    pub fn has_next(&self) -> bool {
        !self.stack.is_empty()
    }

    fn init_sqlite_parameters_table(&mut self) {
        if self.initialized_sqlite_parameters_table {
            return;
        }
        unsafe {
            let original_writable = 0;
            let original_defensive = 0;
            let db = self.connection.db();
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, -1, &original_defensive);
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, 0, 0);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, -1, &original_writable);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, 1, 0);
            let result = self.connection.prepare("CREATE TABLE IF NOT EXISTS temp.sqlite_parameters(key TEXT PRIMARY KEY, value) WITHOUT ROWID");
            result.unwrap().1.expect("TODO").execute().unwrap();
            sqlite3_db_config(db, SQLITE_DBCONFIG_DEFENSIVE, original_defensive, 0);
            sqlite3_db_config(db, SQLITE_DBCONFIG_WRITABLE_SCHEMA, original_writable, 0);
        }
        self.initialized_sqlite_parameters_table = true;
        //self.connection.prepare("CREATE TABLE TEMP.sqlite_parameters(key text, value any ) WITHOUT ROWID")
    }

    pub fn define_parameter(&mut self, key: String, value: String) -> Result<(), String> {
        self.init_sqlite_parameters_table();
        let stmt = self
            .connection
            .prepare("INSERT OR REPLACE INTO temp.sqlite_parameters(key, value) VALUES (?1, ?2)")
            .unwrap()
            .1
            .unwrap();
        stmt.bind_text(1, key);
        stmt.bind_text(2, value);
        stmt.execute().unwrap();
        Ok(())
    }

    pub fn prepare_with_parameters(
        &self,
        sql: &str,
    ) -> Result<(Option<usize>, Option<Statement>), SQLiteError> {
        let (rest, stmt) = self.connection.prepare(sql)?;
        if let Some(stmt) = stmt {
            let params = stmt.bind_parameters();
            for (idx, param) in params.iter().enumerate() {
                let param = if let Some(param) = param.strip_prefix(':') {
                    param
                } else if let Some(param) = param.strip_prefix('@') {
                    param
                } else {
                    param
                };
                match self.lookup_parameter(param) {
                    Some(OwnedValue::Text(s)) => {
                        stmt.bind_text((idx + 1) as i32, std::str::from_utf8(&s).unwrap())
                    }
                    Some(OwnedValue::Integer(v)) => stmt.bind_int64((idx + 1) as i32, v),
                    Some(OwnedValue::Double(v)) => stmt.bind_double((idx + 1) as i32, v),
                    Some(OwnedValue::Blob(v)) => stmt.bind_blob((idx + 1) as i32, v.as_ref()),
                    Some(OwnedValue::Null) => stmt.bind_null((idx + 1) as i32),

                    None => (),
                }
            }
            Ok((rest, Some(stmt)))
        } else {
            Ok((rest, stmt))
        }
    }
    fn lookup_parameter<S: AsRef<str>>(&self, key: S) -> Option<OwnedValue> {
        let stmt = self
            .connection
            .prepare("SELECT value FROM temp.sqlite_parameters WHERE key = ?1")
            .ok()?
            .1?;
        stmt.bind_text(1, key);
        stmt.next()
            .unwrap()
            .map(|v| OwnedValue::from_value_ref(v.get(0).unwrap()))
    }
}

pub fn advance_through_ignorable(contents: &str) -> &str {
    let mut chars = contents.char_indices();
    let mut latest = 0;

    while let Some((idx, c)) = chars.next() {
        latest = idx;
        if c.is_whitespace() {
            continue;
        }
        if c == '-' {
            if let Some((_, '-')) = chars.next() {
                loop {
                    match chars.next() {
                        Some((idx, '\n')) => {
                            latest = idx;
                            break;
                        }
                        Some((_, _)) => continue,
                        None => break,
                    }
                }
            } else {
                break;
            }
        } else if c == '/' {
            if let Some((_idx, '*')) = chars.next() {
                loop {
                    match chars.next() {
                        Some((idx, '*')) => match chars.next() {
                            Some((idx, '/')) => {
                                latest = idx;
                                break;
                            }
                            Some((idx, _)) => {
                                latest = idx;
                                continue;
                            }
                            None => {
                                latest = idx;
                                break;
                            }
                        },
                        Some((idx, _)) => {
                            latest = idx;
                            continue;
                        }
                        None => break,
                    }
                }
            } else {
                latest = idx;
                break;
            }
        } else if c == '#' {
            loop {
                match chars.next() {
                    Some((idx, '\n')) => {
                        latest = idx;
                        break;
                    }
                    Some((_, _)) => continue,
                    None => break,
                }
            }
        } else {
            break;
        }
    }
    contents.get(latest..).unwrap()
}
#[cfg(test)]
mod tests {
    use insta::{assert_debug_snapshot, assert_yaml_snapshot};
    use solite_stdlib::BUILTIN_FUNCTIONS;

    use super::*;
    use crate::{dot::DotCommand, sqlite::Connection};

    #[test]
    fn test_advance_through_ignorable() {
        assert_eq!(advance_through_ignorable("4"), "4");
        assert_eq!(advance_through_ignorable("    4"), "4");

        assert_eq!(advance_through_ignorable("--\n4"), "4");
        assert_eq!(advance_through_ignorable("-- skip me\n4"), "4");
        assert_eq!(
            advance_through_ignorable("-- skip me\n  -- 2nd line \n 4"),
            "4"
        );
        assert_eq!(advance_through_ignorable("-\n4"), "-\n4");

        assert_eq!(advance_through_ignorable("#\n4"), "4");
        assert_eq!(advance_through_ignorable("# skip me\n4"), "4");
        assert_eq!(advance_through_ignorable("# skip me\n  #\n 4"), "4");
        assert_eq!(advance_through_ignorable("#\n--\n4"), "4");

        assert_eq!(advance_through_ignorable("/**/4"), "4");
        assert_eq!(advance_through_ignorable("/* skip me */ 4"), "4");
        assert_eq!(advance_through_ignorable("/** skip me */ 4"), "4");
        assert_eq!(advance_through_ignorable("/* skip me **/ 4"), "4");
    }

    fn functions_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_function_list order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut fns = vec![];
        while let Ok(Some(row)) = stmt.next() {
            fns.push(row.get(0).unwrap().as_str().to_string());
        }
        fns
    }
    fn version_functions_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_function_list where name like '%_version' order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut fns = vec![];
        while let Ok(Some(row)) = stmt.next() {
            fns.push(row.get(0).unwrap().as_str().to_string());
        }
        let mut sql = String::new();
        sql += "select ";
        sql += &fns
            .iter()
            .map(|v| format!("{v}()"))
            .collect::<Vec<String>>()
            .join(", ");
        let stmt = db.prepare(&sql).unwrap().1.unwrap();
        let row = stmt.next().unwrap().unwrap();
        let vers = row.iter().map(|v| v.as_str().to_string());
        assert_eq!(fns.len(), vers.len());
        //fns.iter().zip(vers).collect()
        fns.iter()
            .zip(vers)
            .map(|(func, version)| format!("{func} == {version}"))
            .collect()
    }
    fn modules_of(db: &Connection) -> Vec<String> {
        let stmt = db
            .prepare("select distinct name from pragma_module_list order by 1")
            .unwrap()
            .1
            .unwrap();
        let mut mods = vec![];
        while let Ok(Some(row)) = stmt.next() {
            mods.push(row.first().unwrap().as_str().to_string());
        }
        mods
    }

    #[test]
    fn core_basic() {
        let runtime = Runtime::new(None);
        let stmt = runtime
            .connection
            .prepare("select sqlite_version();")
            .unwrap()
            .1
            .unwrap();
        assert_eq!(
            stmt.next().unwrap().unwrap().get(0).unwrap().as_str(),
            "3.50.1"
        );
        insta::assert_yaml_snapshot!(functions_of(&runtime.connection));
        insta::assert_yaml_snapshot!(modules_of(&runtime.connection));
        insta::assert_yaml_snapshot!(version_functions_of(&runtime.connection));
        insta::assert_yaml_snapshot!(BUILTIN_FUNCTIONS);
    }
    
    #[test]
    fn snap2() {
        let mut rt = Runtime::new(None);
        rt.enqueue(
            "[input]",
            "
-- preamble1
select 1;
select 2;
-- another preamble
select 3.1;select 3.2;
/* inline! */ select 4;

-- what!
select not_exist();",
            BlockSource::File(PathBuf::new()),
        );
        let mut idx = 0;
        loop {
            let step = rt.next_stepx();
            assert_yaml_snapshot!(format!("step-{idx}"), step);
            idx += 1;
            match step {
                None => break,
                _ => (),
            }
        }
    }
}
