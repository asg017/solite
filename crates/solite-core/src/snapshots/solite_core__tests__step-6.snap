---
source: crates/solite-core/src/lib.rs
expression: step
snapshot_kind: text
---
~
